### PR TITLE
framework: Create aspect2d child nodes for window corners & edges

### PR DESCRIPTION
**Pull request by maxrdz**
## Issue description
This change implements the equivalent of the Python ShowBase's aspect2d child nodes that represent the corners and edges of the window in aspect2d coordinates, for the C++ interface of Panda. These nodes allow the user to create child PGui objects relative to a corner or edge of the window under aspect2d. In most applications, the user interface lays on the corners or edges of the screen, so this is a basic feature that C++ developers should also have access to.
This resolves issue #1508.

## Solution description
This PR creates the a2dxxx nodes when aspect2d is requested by the user at `GraphicsWindow::get_aspect_2d`, and updates the nodes' corresponding coordinates on window resize according to the window aspect ratio if aspect2d exists.

## Checklist
I have done my best to ensure that…
* [x] …I have familiarized myself with the CONTRIBUTING.md file
* [x] …this change follows the coding style and design patterns of the codebase
* [x] …I own the intellectual property rights to this code
* [x] …the intent of this change is clearly explained
* [x] …existing uses of the Panda3D API are not broken
* [x] …the changed code is adequately covered by the test suite, where possible.

### CODE REVIEW

**Comment by Max-Rodriguez** [comment]
Feel free to suggest changes. Also let me know if there is a cleaner way to get these nodes rather than having getters for every single node.

**Comment by Max-Rodriguez** [comment]
I ran pytest on the wheel generated from my build. 799 passed, 1 warning, 1 failed. tests/display/test_cg_shader.py failed with the following error: `:shader(error): Tried to load Cg shader, but no Cg support is enabled.` I'm assuming this error is not related to my change?

**Review by rdb — CHANGES_REQUESTED** [review]
Looks good overall

**Comment by rdb** [comment]
> Feel free to suggest changes. Also let me know if there is a cleaner way to get these nodes rather than having getters for every single node.

A single getter with an anchor enum also works but I have no preference here

File: panda/src/framework/windowFramework.cxx [hunk 1]
```diff
@@ -323,6 +323,31 @@ get_aspect_2d() {
...
+    _a2d_right_center = _aspect_2d.attach_new_node("_a2d_right_center");
+    _a2d_top_left = _aspect_2d.attach_new_node("_a2d_top_left");
+    _a2d_top_right = _aspect_2d.attach_new_node("_a2d_top_right");
+    _a2d_bottom_left = _aspect_2d.attach_new_node("_a2d_bottom_left");
+    _a2d_bottom_right = _aspect_2d.attach_new_node("_a2d_bottom_right");
+
+    float a2d_top = 1.0f;
+    float a2d_bottom = -1.0f;
+    float a2d_left = this_aspect_ratio * -1.0f;
+    float a2d_right = this_aspect_ratio;
```
Comment (rdb) [inline panda/src/framework/windowFramework.cxx:340]: `PN_stdfloat`

File: panda/src/framework/windowFramework.cxx [hunk 2]
```diff
@@ -323,6 +323,31 @@ get_aspect_2d() {
...
+
+    // Create child nodes of aspect_2d anchored to edges/corners of the window.
+    _a2d_top_center = _aspect_2d.attach_new_node("a2d_top_center");
+    _a2d_bottom_center = _aspect_2d.attach_new_node("_a2d_bottom_center");
+    _a2d_left_center = _aspect_2d.attach_new_node("_a2d_left_center");
+    _a2d_right_center = _aspect_2d.attach_new_node("_a2d_right_center");
+    _a2d_top_left = _aspect_2d.attach_new_node("_a2d_top_left");
+    _a2d_top_right = _aspect_2d.attach_new_node("_a2d_top_right");
+    _a2d_bottom_left = _aspect_2d.attach_new_node("_a2d_bottom_left");
+    _a2d_bottom_right = _aspect_2d.attach_new_node("_a2d_bottom_right");
```
Comment (rdb) [inline panda/src/framework/windowFramework.cxx:335]: Some of these node names have _ suffix some don't

Let's just give them the same names as in ShowBase, so a2dTopCenter etc, so:
```python
_a2d_bottom_right = _aspect_2d.attach_new_node("a2dBottomRight");
```

File: panda/src/framework/windowFramework.cxx [hunk 3]
```diff
@@ -816,7 +937,38 @@ adjust_dimensions() {
   }
 
   if (!_aspect_2d.is_empty()) {
-    _aspect_2d.set_scale(1.0f / this_aspect_ratio, 1.0f, 1.0f);
+    float a2d_top;
+    float a2d_bottom;
+    float a2d_left;
+    float a2d_right;
```
Comment (rdb) [inline panda/src/framework/windowFramework.cxx:943]: PN_stdfloat

File: panda/src/framework/windowFramework.h [hunk 4]
```diff
@@ -84,6 +84,15 @@ class EXPCL_FRAMEWORK WindowFramework : public TypedWritableReferenceCount {
...
   NodePath get_button_thrower();
 
+  NodePath get_a2d_top_center();
+  NodePath get_a2d_bottom_center();
+  NodePath get_a2d_left_center();
+  NodePath get_a2d_right_center();
+  NodePath get_a2d_top_left();
+  NodePath get_a2d_top_right();
+  NodePath get_a2d_bottom_left();
+  NodePath get_a2d_bottom_right();
```
Comment (rdb) [inline panda/src/framework/windowFramework.h:94]: Let's make these functions INLINE (mark them so both the implementation and the declaration) and move them to the .I file.

**Review by rdb — CHANGES_REQUESTED** [review]
Looks good, almost there

File: panda/src/framework/windowFramework.h (hunk 4 above)
Comment (Max-Rodriguez) [reply panda/src/framework/windowFramework.h:94]: Nevermind, looked up more info on inline :)

**Comment by Max-Rodriguez** [comment]
Cleaned up commit history. Removed merge commit, rebased, and squashed requested changes into one commit.

**Comment by Moguri** [comment]
I went ahead and updated (merged in master) the branch to try and get the suck CI jobs running. Feel free to rebase if you'd prefer.

**Review by rdb — APPROVED** [review]
Thanks!

**Pull request closed without merging**